src/Sylius/Bundle/TaxonomyBundle/Doctrine/ORM/TaxonRepository.php

```diff
@@ -92,13 +92,20 @@ public function findRootNodes(): array
 
     public function findByNamePart(string $phrase, ?string $locale = null, ?int $limit = null): array
     {
-        return $this->createTranslationBasedQueryBuilder($locale)
+        /** @var TaxonInterface[] $results */
+        $results = $this->createTranslationBasedQueryBuilder($locale)
             ->andWhere('translation.name LIKE :name')
             ->setParameter('name', '%' . $phrase . '%')
             ->setMaxResults($limit)
             ->getQuery()
             ->getResult()
         ;
+
+        foreach ($results as $result) {
+            $result->setFallbackLocale(array_key_first($result->getTranslations()->toArray()));
```

Comment:
And since this is a collection, there is no option to get the first value right away? the function `first ()` should get the hang of the subject. The second point is casting the collection to an array and extracting one element is disputable. It would add a kinda big overhead to the search especially when it is foreached.

Reply:
The `key` method can be used: `$result->getTranslations()->key()`, but this will modify the array pointer.
However, I don't understand where does it cause a _big overhead_, can you elaborate?
The `->toArray()` call just returns the underlying array structure and `array_key_first` is the most efficient way to do the job without modifying the pointer, according to the [RFC ](https://wiki.php.net/rfc/array_key_first_last_index) and the [docs](https://www.php.net/manual/en/function.array-key-first.php).

Reply:
I also don't see any big overhead here 🤔

Reply:
I just created a super answer and removed it ehh... Long story short.

The `$result->getTranslations()->toArray()` is costly when it comes to memory, because under `$result->getTranslations()` is Doctrine Collection and as far as I remember the `toArray()` function will create copy of this array and don't get me wrong there is no fault in `array_key_first` function, yet `$result->getTranslations()->first()` will be simpler and faster approach. Yet I was wrong, because then I looked at the https://github.com/doctrine/collections/blob/2.0.x/lib/Doctrine/Common/Collections/ArrayCollection.php and see that first and key actually are modifying the array.

Reply:
Since you looked at the implementation, take a look at `toArray()` method too, because it just returns the underlying array.
Yes, that will perform a shallow copy of the **one element array**, which is far from being a **big overhead**.

https://github.com/doctrine/collections/blob/c1ec9823f9250fefb274cccf62970b5c167e6bc7/lib/Doctrine/Common/Collections/ArrayCollection.php#L69-L72